### ISSUE TITLE
ipf is a no-op

### ISSUE THREAD

**Issue by Ghoughpteighbteau** [colobot/colobot#410]
I have a bot that made use of classes to simplify the code, but this has put the code well past the 100 instructions per frame limit and has slowed it down. 

Normally I'd be ok with having a cost for over complicating code, but seeing as I got bit for modularizing my code, that feels unfair :(

A quick waltz through the code suggests the m_ipf is in fact getting set, but only after the script has begun to run, and since that property is never queried again, it never has the opportunity to be anything other than 100 instructions per frame. 

ipf is only used in one instance, ai/lady02.txt

**Comment by attila123** [comment]
Is the fix/implementation for this difficult? Would be nice to have it in the 0.1.6, because it affects gameplay: #611 .

**Comment by tomangelo2** [comment]
CBOT is going to be refactored at 0.2.0, so all noncritical cbot related bugs are moved to this release. Before #611 it wasn't regadred as important. That's why this issue is waiting so long to fix. But in this case we'll look if there will be simple fix for that.

**Comment by attila123** [comment]
Maybe one possible solution until ipf is corrected is to modify the AlienQueen's script so that it does not use ipf. Didn't check whether that is feasible though.

**Comment by tomangelo2** [comment]
We should try that at first place, but in case we couldn't do that, we can try change variable
``` c++
const int CBOT_IPF = 100;       // CBOT: default number of instructions / frame
```
to enough to provide correct behaviour of AlienQueen's script.

**Comment by krzys-h** [comment]
CBOT_IPF is only a default value for m_ipf variable which the one that is actually used (or at least it looked like it should when I last looked at this issue). We can't just change the default to the value AlienQueen uses (10 I believe) because that would slow down all CBot scripts 10 times, and they are too slow for some calculations already.

**Comment by Ghoughpteighbteau** [comment]
lol, if I recal the IPF for the queen was set to 2. So yah. the queen will have to have some waits inserted if you want to slow it down without fixing IPF. lady02.txt isn't that complicated though...

**Comment by attila123** [comment]
For me ipf() smells like a huge anti-pattern, and I think the best to do with that is to remove that!

At the same time maybe introduce something like
```
doForMillis( time ) {
  // code
}
```
. It would execute the code in the block and if that took less than 'time', it would wait( 'time' - <code execution time>). <code execution time> can be significant in case it contains instructions like goto().

**Comment by attila123** [comment]
I can imagine that ipf mechanism was originally added for protection against possible busy loops in bot scripts and later abused to set the pace for certain programs like in the AlienQueen's script.
What I don't like about ipf() is that if you add additional instructions to the code scheduled by ipf(), e.g. just to do some better calculations for certain movements, etc., it will slow it down. Well, for certain situations it's not relevant though.
So for me ipf() seems to be the necessary bad, because you can't trust the users to use wait() in the bot scripts from time to time, e.g. in loops.

**Comment by Ghoughpteighbteau** [comment]
well. I mean you absolutely have to have IPF. I mean, the program can't crash because someone got into an infinite loop. In a programming game with a bunch of novices. Yikes. 

I kind of agree with you, IPF should be removed. Processor speed should be a gameplay mechanic if anything, but I don't know if colobot can include something like that, I think it tries to keep things simple (lame :P).

So IPF is kind of necessary if you want to try to do more complicated stuff and not have it perform terribly.

**Comment by krzys-h** [comment]
This is why ipf() is a hidden and undocumented feature. You are not supposed to use it unless you really are the master of this game and the limit really bothers you.
@MrSimbax add it to the cheats documentation page, please?

**Comment by MrSimbax** [comment]
@krzys-h Done.

**Comment by hexagonrecursion** [comment]
`ipf()` is not a no-op:

```c++
extern void object::New()
{
	float s = abstime();
	int i = 0;
	while(s == abstime()) {
		++i;
	}
	message(i); // prints 12
	
	ipf(1000);
	wait(0);
	s = abstime();
	i = 0;
	while(s == abstime()) {
		++i;
	}
	message(i);  // prints 124
}
```
